presto-main/src/main/java/com/facebook/presto/operator/DriverStatus.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public class DriverStatus
+{
+    private final ScheduledExecutorService executor;
+
+    private long runningSequence;
+    private boolean running;
+    private final AtomicBoolean yield = new AtomicBoolean();
+
+    public DriverStatus(ScheduledExecutorService executor)
+    {
+        this.executor = executor;
+    }
+
+    public synchronized void run(long maxRunNanos)
+    {
+        checkArgument(!running, "driver was running");
+        checkArgument(!shouldYield(), "yield while driver was not running");
+        this.runningSequence++;
+        long runningSequence = this.runningSequence;
+        executor.schedule(() -> {
+            synchronized (this) {
+                if (runningSequence == this.runningSequence && this.running) {
+                    yield.set(true);
+                }
+            }
+        }, maxRunNanos, NANOSECONDS);
+        running = true;
+    }
```

Comment:
We talked offline about restructuring this class so that it is easier to use in a test.